### DISCUSSION TITLE
Is there any other way to get the error message if authentication fails in Service Classes?

### DISCUSSION THREAD

**Discussion by morcef** [CrowdStrike/falconpy#500] [Q&A] [answered]
Hi,
I am new to this SDK, and I have been starting with using the Detects Service Class (https://falconpy.io/Service-Collections/Detects.html). In the beginning the Detects authenticated() method would return False, and I was having trouble finding the reason. 
This seems to be due to the fact that the Service Class will throw away any response that does not have status code 201, and include a token (https://github.com/CrowdStrike/falconpy/blob/5adb86cab114423e4959ecbed3ae49a97194ae5a/src/falconpy/_service_class.py#L142-L156). 
At this point there was no clear way for me to figure out why this was failing, other than accessing the Detects' auth_object, and executing its token() method a second time to get the actual response. However, my IDE is giving me the following warning when accessing the auth_object: Unresolved attribute reference 'auth_object' for class 'Detects'.

**Comment by jshcodes** [comment, marked as answer]
Can I see a sample of your code?  You're correct that you should be able to call token() straight from the auth_object attached to the service class instance.

> Please remember to redact any API credentials, CID values or AID values.

**Reply by jshcodes** [reply]
As a follow up, I have created [Issue 501](https://github.com/CrowdStrike/falconpy/issues/501). With some minor adjustments, we should be able to provide you the failure reason as an attribute on the Service Class.

**Reply by morcef** [reply]
Thank you for your quick reply. I was also thinking that this might be an issue to improve the Service Class with a proper "failure reason" attribute, but wanted to discuss any alternatives first.
I am not really sure my code will provide any further details, but this is a basic example:
```
from falconpy import Detects

detects = Detects(client_id="client_id", client_secret="client_secret")
if not detects.authenticated():
    response = detects.auth_object.token()
    print(f"{response.get('status_code')}: {response.get('body', {}).get('errors', [])[0]['message']}")
```

In this case my IDE (Pycharm) will give this warning: Unresolved attribute reference 'auth_object' for class 'Detects' :5

**Reply by jshcodes** [reply]
_Interesting_.  I've been using VSCode recently, and don't get an alert for this.  I'll fire up PyCharm and see if I can't run down the cause of the IDE warning.  (Your code above _does_ run for me and produces our error message properly.)

... but that's silly, you shouldn't have to run a second request for this.  I'm in the process of unit testing v0.9.0 now, which contains an update providing new attributes on `OAuth2` and `APIHarness` to speak to issue #501. Your new pattern will be:

```python
from falconpy import Detects

detects = Detects(client_id="client_id", client_secret="client_secret")
if not detects.authenticated():
    print(detects.token_fail_reason)
```

At the moment, I don't have it passing the error code back (either 400 or 403 depending on the failure variation), but if we think it's valuable it's not a difficult add.

**Reply by jshcodes** [reply]
Played around with this some more and decided to pass back the token error code as well.

```python
from falconpy import Detects

detects = Detects(client_id="client_id", client_secret="client_secret")
if detects.token_status != 201:
    print(detects.token_fail_reason)
```

**Reply by morcef** [reply]
Great! 
Am I correct when assuming that for any RequestExceptions that are not HTTPErrors, i.e. ConnectionError or Timeout, the token_status will be 500 (default error code), and the token_fail_reason will be the exception message?

**Reply by jshcodes** [reply]
I'll need to test with an example scenario to create a 500. I tested generating 400's, 403's and 415's and received the appropriate error response. 

> These attribute values will only change when:
>    - making a call to the `token` method
>    - instantiating an instance of a Service Class (which calls the `token` method)
>    - making a call that auto-refreshes your token
>    - when calling the `authenticate` method in the Uber Class. (If you don't call `authenticate` and call `command` first, it calls `authenticate` for you.)

**Reply by jshcodes** [reply]
Confirmed. If you get a Timeout / ConnectionError response back, it's a 500 and populates properly.

**Reply by jshcodes** [reply]
Status update: v0.9.0 has passed unit testing, merged to main and is now available on the test package index. ([Bleeding edge release announcement](https://github.com/CrowdStrike/falconpy/discussions/505)) This update will roll to the production index tomorrow.

#### Installing the bleeding-edge release
You can check out this new release by installing the package via the Python Package _Test_ Index.
```shell
python3 -m pip install -i https://test.pypi.org/simple/ crowdstrike-falconpy
```
or
```shell
pip3 install -i https://test.pypi.org/simple/ crowdstrike-falconpy
```

#### Upgrading to the bleeding-edge release
You can upgrade to this new release by installing the update via the Python Package _Test_ Index.
```shell
python3 -m pip install -i https://test.pypi.org/simple/ crowdstrike-falconpy --upgrade
```
or
```shell
pip3 install -i https://test.pypi.org/simple/ crowdstrike-falconpy --upgrade
```

**Reply by morcef** [reply]
Thank you very much!

**Reply by morcef** [reply]
> _Interesting_. I've been using VSCode recently, and don't get an alert for this. I'll fire up PyCharm and see if I can't run down the cause of the IDE warning. (Your code above _does_ run for me and produces our error message properly.)

Btw, did you figure out about the PyCharm warning?

**Reply by jshcodes** [reply]
Yes I have! Syntactically, PyCharm is annoyed that auth_object is not defined explicitly within _service_class.py. (It's value is set based upon login variation, which is the same time it is defined.)  It does not appear to prevent code function in any scenario I've tested, but the highlight in PyCharm is super-annoying.

I'm working on an update that should resolve this warning now. (Will release as part of the v0.9.0 code base.) :smile:

**Reply by morcef** [reply]
>the highlight in PyCharm is super-annoying

Haha, it most definitely is!

>PyCharm is annoyed that auth_object is not defined explicitly within _service_class.py. (It's value is set based upon login variation, which is the same time it is defined.)

That's what I was worried was the cause as well. Thank you for looking into that as well.

**Reply by jshcodes** [reply]
Absolutely not a problem. Thank you for reporting these, they are both excellent contributions! :smile:

**Reply by morcef** [reply]
Hi @jshcodes :) 

I just wanted to inform you that I am still experiencing the annoying PyCharm warnings.

```
auth = OAuth2(client_id="client_id", client_secret="client_secret")

auth.token()

if not auth.authenticated():
    print(f"Status code: {auth.token_status}, message: {auth.token_fail_reason}")
```

This code works fine, but still gives me:
```
Unresolved attribute reference 'authenticated' for class 'OAuth2'
Unresolved attribute reference 'token_status' for class 'OAuth2'
Unresolved attribute reference 'token_fail_reason' for class 'OAuth2'
```

**Reply by jshcodes** [reply]
Booooo JetBrains!  😞 

Let's handle this as a [new issue](https://github.com/CrowdStrike/falconpy/issues/554) and get that resolved.

**Reply by jshcodes** [reply]
Hi @morcef!

I am unable to recreate the issue as described above. (See ticket [here](https://github.com/CrowdStrike/falconpy/issues/554).)  Is this still occurring?
